src/main/java/net/neoforged/neoforge/fluids/capability/wrappers/CauldronWrapper.java

```diff
@@ -0,0 +1,139 @@
+    }
+
+    private CauldronFluidContent getContent(BlockState state) {
+        CauldronFluidContent content = CauldronFluidContent.getForBlock(state.getBlock());
+        if (content == null) {
+            throw new IllegalStateException("Unexpected error: no cauldron at location " + pos);
+        }
+        return content;
+    }
+
+    @Override
+    public FluidStack getFluidInTank(int tank) {
+        BlockState state = level.getBlockState(pos);
+        CauldronFluidContent contents = getContent(state);
+        return new FluidStack(contents.fluid, contents.totalAmount * contents.currentLevel(state) / contents.maxLevel);
+    }
+
+    @Override
+    public int getTankCapacity(int tank) {
+        BlockState state = level.getBlockState(pos);
+        CauldronFluidContent contents = getContent(state);
+        return contents.totalAmount;
+    }
+
+    @Override
+    public boolean isFluidValid(int tank, FluidStack stack) {
+        return CauldronFluidContent.getForFluid(stack.getFluid()) != null;
+    }
+
+    // Called by fill and drain to update the block state.
+    private void updateLevel(CauldronFluidContent newContent, int level, FluidAction action) {
+        if (action.execute()) {
+            BlockState newState = newContent.block.defaultBlockState();
+
+            if (newContent.levelProperty != null) {
+                newState = newState.setValue(newContent.levelProperty, level);
+            }
+
+            this.level.setBlockAndUpdate(pos, newState);
+        }
+    }
+
+    @Override
+    public int fill(FluidStack resource, FluidAction action) {
+        if (resource.isEmpty()) {
+            return 0;
+        }
+
+        CauldronFluidContent insertContent = CauldronFluidContent.getForFluid(resource.getFluid());
+        if (insertContent == null) {
+            return 0;
+        }
+
+        BlockState state = level.getBlockState(pos);
+        CauldronFluidContent currentContent = getContent(state);
+        if (currentContent.fluid != Fluids.EMPTY && currentContent.fluid != resource.getFluid()) {
+            // Fluid mismatch
+            return 0;
+        }
+
+        // We can only insert increments based on the GCD between the number of levels and the total amount.
+        int d = IntMath.gcd(insertContent.maxLevel, insertContent.totalAmount);
+        int amountIncrements = insertContent.totalAmount / d;
+        int levelIncrements = insertContent.maxLevel / d;
+
+        int insertedIncrements = Math.min(resource.getAmount() / amountIncrements, (insertContent.maxLevel - currentContent.currentLevel(state)) / levelIncrements);
+        if (insertedIncrements > 0) {
+            updateLevel(insertContent, currentContent.currentLevel(state) + insertedIncrements * levelIncrements, action);
+        }
+
+        return insertedIncrements * amountIncrements;
+    }
+
+    @Override
+    public FluidStack drain(FluidStack resource, FluidAction action) {
+        FluidStack current = getFluidInTank(0);
```

Comment:
We should add a helper private drain method that can take the `BlockState` as both `getFluidInTank` and the later `drain` method that are called by this `drain` implementation query the state from the world... which is not particularly great...